app/controllers/auth/AuthAction.scala

```diff
@@ -0,0 +1,135 @@
+import play.api.{Configuration, Play}
+import uk.gov.hmrc.auth.core.{AuthorisedFunctions, ConfidenceLevel, Enrolment, Enrolments, InsufficientConfidenceLevel, InsufficientEnrolments, NoActiveSession, PlayAuthConnector}
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.domain.{CtUtr, EmpRef, Nino, SaUtr, Uar, Vrn}
+import uk.gov.hmrc.http.{CorePost, HeaderCarrier}
+import uk.gov.hmrc.play.HeaderCarrierConverter
+import uk.gov.hmrc.play.config.ServicesConfig
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthActionImpl @Inject()(override val authConnector: AuthConnector,
+                               configuration: Configuration)(implicit ec: ExecutionContext)
+  extends AuthAction with AuthorisedFunctions {
+
+  override def invokeBlock[A](request: Request[A], block: AuthenticatedRequest[A] => Future[Result]): Future[Result] = {
+
+    implicit val hc: HeaderCarrier =
+      HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+
+    authorised(ConfidenceLevel.L50 and (Enrolment("IR-SA") or Enrolment("IR-SA-AGENT")))
+      .retrieve(Retrievals.allEnrolments and Retrievals.externalId and Retrievals.nino) {
+        case Enrolments(enrolments) ~ Some(externalId) ~ nino => {
+          val agentRef: Option[Uar] = enrolments.find(_.key == "IR-SA-AGENT").flatMap { enrolment =>
+            enrolment.identifiers
+              .find(id => id.key == "IRAgentReference")
+              .map(key => Uar(key.value))
+          }
+          val saUtr: Option[SaUtr] = enrolments.find(_.key == "IR-SA").flatMap { enrolment =>
+            enrolment.identifiers
+              .find(id => id.key == "UTR")
+              .map(key => SaUtr(key.value))
+          }
+
+          val payeEmpRef: Option[EmpRef] = enrolments.find(_.key == "IR-PAYE")
+            .map { enrolment =>
+              val taxOfficeNumber = enrolment.identifiers.find(id => id.key == "TaxOfficeNumber").map(_.value)
+              val taxOfficeReference = enrolment.identifiers.find(id => id.key == "TaxOfficeReference").map(_.value)
+
+              (taxOfficeNumber, taxOfficeReference) match {
+                case (Some(number), Some(reference)) =>
+                  EmpRef(number, reference)
+              }
+            }
+
+          val ctUtr: Option[CtUtr] = enrolments.find(_.key == "IR-CT").flatMap { enrolment =>
+            enrolment.identifiers
+              .find(id => id.key == "UTR")
+              .map(key => CtUtr(key.value))
+          }
+
+          val vrn: Option[Vrn] = enrolments.find(vatEnrolments => vatEnrolments.key == "HMCE-VATDEC-ORG" || vatEnrolments.key == "HMCE-VATVAR-ORG")
+            .flatMap { enrolment =>
+              enrolment.identifiers
+                .find(id => id.key == "VATRegNo")
+                .map(key => Vrn(key.value))
+            }
+
+          block {
+            AuthenticatedRequest(
+              externalId,
```

Comment:
field in the analytics is userId, why are we using the externalId? Might be worth looking at the splunk with the analysts to see what that field looks like

Reply:
No UserId available from retrievals, Confluence guide said to use either InternalId or ExternalId to replace UserId, both of which are just a random number.

Reply:
Ok, great